### DISCUSSION TITLE
15. Adding more coverage to the tests

### DISCUSSION THREAD

**Discussion by BinayakBanerjee** [Cyfrin/foundry-full-course-cu#2154] [General]
Can someone explain to me the following lines of code ? I have commented out the lines where I have doubt.

    function testOnlyOwnerCanWithdraw() public funded {
            vm.prank(USER);
            fundMe.fund{value: SEND_VALUE}(); 
    
            vm.expectRevert(); // Why is this put, what is the cause ? 
            vm.prank(USER); // Why is vm.prank done again when it was just done above, does vm.prank sets user only for the next line of code ?
            fundMe.withdraw();
        }

**Comment by EngrPips** [comment]
```js
function testOnlyOwnerCanWithdraw() public funded {
// we prank the `USER`, this basically tells foundry that we want the next transaction to be sent by the `USER`
        vm.prank(USER);
// we fund the `FundMe` contract by sending some ether to the contract, because we pranked `USER` earlier so this transaction is sent by `USER`
        fundMe.fund{value: SEND_VALUE}(); 
// we tell foundry that the next transaction that will happen will revert by using `vm.expectRevert()` cheat code 
        vm.expectRevert(); // Why is this put, what is the cause ? 
// we prank `USER`, I believe you know prank does now I explained earlier
        vm.prank(USER); // Why is vm.prank done again when it was just done above, does vm.prank sets user only for the next line of code ?
// we call the withdraw function on the `FundMe` contract, `USER` is the one that sent this transaction but `USER` is not the owner of the `FundMe` contract so this transaction will revert because the `FundMe` contract only expects it's owner to withdraw from it.
        fundMe.withdraw();
    }
```
If anything is still confusing, Please ask questions.

**Reply by BinayakBanerjee** [reply]
Hello @EngrPips , thanks for the detailed explanation. While I'm now clear that **vm.prank(USER);** works only for the subsequent line of code, I'm still not clear about the need or usecase of this code with respect to the test case **vm.expectRevert();**

```
// Not clear what is the need to expectRevert ? Is it sort of an exception handling procedure in case withdraw fails ? Will it ever fail in this scenrio ?
          vm.expectRevert();
// Below line is ignored by vm.expectRevert(); I believe and its actually applicable on the withdraw function 
          vm.prank(USER);
// If we are setting the user with vm.prank and then withdrawing from the same user - AKA the contract owner, then when would the 
expectRevert(); be called ? Is this just redundant code in this scenario ? 
          fundMe.withdraw();
```

**Reply by EngrPips** [reply]
To be clear, `vm.prank(USER)` does not work for the next line of code but the next line that sends a transaction in our code, and you know sometimes some things will not make sense until later. Keep going with the course; things will make more sense with time.

**Reply by BinayakBanerjee** [reply]
Right. Didn't know that. Thanks.

**Reply by EngrPips** [reply]
You are welcome. Keep crushing it.
